parallel/slice.go

```diff
@@ -1,6 +1,35 @@
 package parallel
 
-import "sync"
+import (
+	"sync"
+)
+
+// Filter iterates over elements of collection, returning an array of all elements predicate returns truthy for.
+// `predicate` is call in parallel.
+func Filter[V any](collection []V, predicate func(V, int) bool) []V {
+	result := []V{}
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	wg.Add(len(collection))
```

Comment:
This is not really a smart thing to do as lock and unlock the mutex over and over again. Say we were to have 10 goroutines, each one taking longer than the previous one. Using lop.Filter would actually be slower because we have mutexes to lock, unlock and goroutines to create. 

Smarter would be a pool of workers with channels sending jobs to them. So create `runtime.NumCPU()`s amount of goroutines, send tasks to them, collect all results and proceed. This gives worse best case performance but way better average performance.

Reply:
ok, thanks, so we need lop.Filter? I will try to revise it again.

Reply:
We should put predicate results into a map, key being the item index.

Then we create the final slice to output.

Reply:
About pooling, I'm not even sure it is necessary, because Go already limits the number of concurrent goroutines.

For applications having very high load, pools would need to be shared globally, between every helper calls. So is it the job of this library to handle those cases?

WDYT?

Reply:
Yes, the go runtime handles pooling in theory but consider that for every goroutine a new stack is created. Each goroutine takes minimum 2kb. For 1000 elements this is 2mb. That's absolutely insane. + It will probably be slower than just using a for loop. A faster solution would be to implement a task pool.  I am currently working on a lo feature that implements a task pool. If you can wait a day, you can use that instead :)

Reply:
> 

```go
func Filter[V any](collection []V, predicate func(V, int) bool) []V {
	result := []V{}

	var wg sync.WaitGroup
	var mu sync.Mutex
	wg.Add(len(collection))

	for i, item := range collection {
		go func(_item V, _i int) {
			key := predicate(_item, _i)
			mu.Lock()
			if key {
				result = append(result, _item)
			}

			mu.Unlock()

			wg.Done()
		}(item, i)
	}

	wg.Wait()

	return result
}
```
I don't understand why we need put predicate results into a map. Is it ok?

Reply:
```go
func Filter[V any](collection []V, predicate func(V, int) bool) []V {
        tmp := make([]bool, len(collection))

        // map
	var wg sync.WaitGroup
	wg.Add(len(collection))
	for i, item := range collection {
		go func(_item V, _i int) {
			tmp[_i] = predicate(_item, _i)
			wg.Done()
		}(item, i)
	}
	wg.Wait()

        // reduce
	result := []V{}
	for i, item := range collection {
	        if ok, _ := tmp[i]; ok {
                        result = append(result, item)
                }
        }

	return result
}
```

It looks like a map/reduce jobs.

Reply:
yeah, I think we can refer to https://github.com/kevwan/mapreduce library